caffe2/operators/crop_and_resize_op.cu

```diff
@@ -0,0 +1,393 @@
+            const float bottom = bottom_left + (bottom_right - bottom_left) * x_lerp;
+
+            crops_data[out_idx] = top + (bottom - top) * y_lerp;
+
+        } else {
+
+            const int nearest_x_index = static_cast<int>(in_x + 0.5f);
+            const int nearest_y_index = static_cast<int>(in_y + 0.5f);
+
+            crops_data[out_idx] = static_cast<float>(fm_data[batch_index * channels * fm_height * fm_width +
+                c * fm_height * fm_width +
+                nearest_y_index * fm_width +
+                nearest_x_index]);
+
+        }
+
+    }
+
+}
+
+template<typename T>
+__global__ void CropAndResizeBackward(
+    const int nthreads,
+    const T* dY_data,
+    const T* boxes_data,
+    int batch_size,
+    int fm_height,
+    int fm_width,
+    int crop_height,
+    int crop_width,
+    int channels,
+    int num_boxes,
+    int box_dim,
+    int method,
+    T* dX_data)
+{
+
+    CUDA_1D_KERNEL_LOOP(out_idx, nthreads) {
+
+        const int index = out_idx;
+
+        const int x = index % crop_width;
+        const int y = (index / crop_width) % crop_height;
+        const int c = (index / crop_width / crop_height) % channels;
+        const int b = index / crop_width / crop_height / channels;
+
+        const int batch_index = static_cast<int>(boxes_data[b * box_dim]);
+
+        if (batch_index < 0 || batch_index >= batch_size) {
+            continue;
+        }
+
+        const float x1 = static_cast<float>(boxes_data[b * box_dim + 1]);
+        const float y1 = static_cast<float>(boxes_data[b * box_dim + 2]);
+        const float x2 = static_cast<float>(boxes_data[b * box_dim + 3]);
+        const float y2 = static_cast<float>(boxes_data[b * box_dim + 4]);
+
+        const float height_scale = (crop_height > 1) ?
+            ((y2 - y1) * (fm_height - 1) / (crop_height - 1)) : 0.0f;
+
+        const float width_scale = (crop_width > 1) ?
+            ((x2 - x1) * (fm_width - 1) / (crop_width - 1)) : 0.0f;
+
+        const float in_y = (crop_height > 1) ?
+            (y1 * (fm_height - 1) + y * height_scale) :
+            (0.5f * (y1 + y2) * (fm_height - 1));
+
+
+        if (in_y < 0 || in_y > fm_height - 1) {
+            continue;
+        }
+
+        const float in_x = (crop_width > 1) ?
+            (x1 * (fm_width - 1) + x * width_scale) :
+            (0.5f * (x1 + x2) * (fm_width - 1));
+
+        if (in_x < 0 || in_x > fm_width - 1) {
+            continue;
+        }
+
+        if (method == BILINEAR) {
+
+            const int top_y_index = floorf(in_y);
+            const int bottom_y_index = ceilf(in_y);
+            const float y_lerp = in_y - top_y_index;
+
+            const int left_x_index = floorf(in_x);
+            const int right_x_index = ceilf(in_x);
+            const float x_lerp = in_x - left_x_index;
+
+            const float dtop = (1 - y_lerp) * dY_data[out_idx];
+
+            gpu_atomic_add(static_cast<T>((1 - x_lerp) * dtop),
```

Comment:
It seems we have to change thread parallelization mechanics to avoid atomic operations when accumulating gradients, as `CUDA_1D_KERNEL_LOOP` asks us to be accurate when adding in the same place from parallel threads.
This implementation is typical for backward pass of operators with semantics "take feature maps and boxes and crop the regions of interests". For example, this is gpu_atomic_add in [roi_align](https://github.com/pytorch/pytorch/blob/6c7135decba2e9090fd663f6d520311a10c84ed8/caffe2/operators/roi_align_gradient_op.cu#L174) (roi align is perhaps semantically closest to crop_and_resize, so I've chosen it).